### PR TITLE
feature(rjy): add mtcar env and muzero visualization

### PR DESCRIPTION
**Pull request by nighood**
This code provides a visualization analysis of the MountainCar environment's observation space and the latent states of a trained MuZero model using PCA (Principal Component Analysis) according to paper [Visualizing MuZero Models](http://arxiv.org/abs/2102.12924). The goal is to gain insights into the specific types of representations learned by these models.

The code performs the following steps:

1. Import the required libraries and modules.
2. Load the MountainCar environment and its observation space.
3. Load a pre-trained MuZero model.
4. Generate random observation samples (observations) from the MountainCar environment.
5. Perform PCA to reduce the dimensionality of the original observation space.
6. Extract the latent states from the MuZero model by inputting the observation samples into the model's representation network and extracting the output.
7. Apply PCA to reduce the dimensionality of the latent states.
8. Visualize the reduced observation space and latent states.

By conducting these visualization analyses, we can gain a deeper understanding of how the MuZero model learns and represents information in the MountainCar environment.

For more information about the mountain_car environment, see [mountain_car doc](https://www.gymlibrary.dev/environments/classic_control/mountain_car/).

Performance of the Muzero model used in the code on mountain car env:
image

### CODE REVIEW

**Comment by puyuan1996** [comment]
增加这个PR功能的 description，以及把可视化相关代码也提交上来吧。代码需要增加详细规则的注释，请参考lunarlander的环境。

**Comment by puyuan1996** [comment]
目前mtcar muzero的性能是如何的呢

File: zoo/classic_control/mountain_car/entry/mountain_car_eval.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
+from zoo.classic_control.mountain_car.config.mtcar_muzero_config import main_config, create_config
```
Comment (puyuan1996) [inline zoo/classic_control/mountain_car/entry/mountain_car_eval.py:1]: 3个ipynb各自的不同，主要功能在开始写清楚哈，可以先中文，但是后面需要转成英文版本的哈

File: zoo/classic_control/mountain_car/config/mtcar_muzero_config.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    if entry_type == "train_muzero":
+        from lzero.entry import train_muzero
+    elif entry_type == "train_muzero_with_gym_env":
+        """
+        The ``train_muzero_with_gym_env`` entry means that the environment used in the training process is generated by wrapping the original gym environment with LightZeroEnvWrapper.
+        Users can refer to lzero/envs/wrappers for more details.
+        """
+        from lzero.entry import train_muzero_with_gym_env as train_muzero
+
+    train_muzero([main_config, create_config], seed=0, max_env_step=max_env_step)
```
Comment (puyuan1996) [inline zoo/classic_control/mountain_car/config/mtcar_muzero_config.py:102]: 把train_muzero_with_gym_env相关的先去掉哈，暂时用不到

**Comment by puyuan1996** [comment]
> 目前mtcar muzero的性能是如何的呢

把目前的基线结果及其相关介绍放在上面的description里面哈

File: zoo/classic_control/mountain_car/config/mtcar_muzero_config.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+update_per_collect = 100
+batch_size = 256
+max_env_step = int(1e6)
+reanalyze_ratio = 0
+# ==============================================================
+# end of the most frequently changed config specified by the user
+# ==============================================================
+
+mountain_car_muzero_config = dict(
+    exp_name=f'lz_result/debug/mountain_car_muzero_ns{num_simulations}_upc{update_per_collect}_rr{reanalyze_ratio}_seed0',
```
Comment (puyuan1996) [inline zoo/classic_control/mountain_car/config/mtcar_muzero_config.py:19]: exp_name按照我们已有config的格式，去掉 lz_result/debug/ 这个前缀

File: zoo/classic_control/mountain_car/entry/mountain_car_eval.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+    returns_mean_seeds = np.array(returns_mean_seeds)
+    returns_seeds = np.array(returns_seeds)
+
+    # Print evaluation results
+    print("=" * 20)
+    print(f"We evaluated a total of {len(seeds)} seeds. For each seed, we evaluated {num_episodes_each_seed} episode(s).")
+    print(f"For seeds {seeds}, the mean returns are {returns_mean_seeds}, and the returns are {returns_seeds}.")
+    print("Across all seeds, the mean reward is:", returns_mean_seeds.mean())
+    print("=" * 20)
```
Comment (puyuan1996) [inline zoo/classic_control/mountain_car/entry/mountain_car_eval.py:56]: 目前只保留zoo/classic_control/mountain_car/entry/vis_mz_mtcar.ipynb这个吧，命名中vis换成完整的 visualize，以及文件中注明参考的论文和链接，写一下整体的介绍

File: lzero/worker/muzero_evaluator.py [hunk 5]
```diff
@@ -199,6 +199,7 @@ def eval(
             train_iter: int = -1,
             envstep: int = -1,
             n_episode: Optional[int] = None,
+            return_trajectory: bool = False,
     ) -> Tuple[bool, float]:
         """
         Overview:
```
Comment (puyuan1996) [inline lzero/worker/muzero_evaluator.py:205]: 关于添加的episode_info['trajectory']，注释中增加相应的说明

File: zoo/classic_control/mountain_car/envs/mtcar_lightzero_env.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        obs = {'observation': obs, 'action_mask': action_mask, 'to_play': -1}
+        return obs
+
+    def step(self, action: np.ndarray) -> BaseEnvTimestep:
+
+        # Making sure that input action is of numpy ndarray
+        # assert isinstance(action, np.ndarray), type(action)
+
+        # Extract action as int, 0-dim array
+        # action = action.squeeze()
```
Comment (puyuan1996) [inline zoo/classic_control/mountain_car/envs/mtcar_lightzero_env.py:85]: 无用的注释删除了

**Review by puyuan1996 — CHANGES_REQUESTED** [review]

File: zoo/classic_control/mountain_car/config/mtcar_muzero_config.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    policy=dict(
+        type='muzero',
+        import_names=['lzero.policy.muzero'],
+    ),
+)
+mountain_car_muzero_create_config = EasyDict(mountain_car_muzero_create_config)
+create_config = mountain_car_muzero_create_config
+
+if __name__ == "__main__":
+    # Users can use different train entry by specifying the entry_type.
```
Comment (puyuan1996) [inline zoo/classic_control/mountain_car/config/mtcar_muzero_config.py:90]: 无用的注释删除了

**Review by puyuan1996 — CHANGES_REQUESTED** [review]

File: lzero/worker/muzero_evaluator.py [hunk 8]
```diff
@@ -455,4 +457,6 @@ def eval(
             stop_flag, episode_info = objects
 
         episode_info = to_item(episode_info)
+        if return_trajectory:
+            episode_info['trajectory'] = game_segments
```
Comment (puyuan1996) [inline lzero/worker/muzero_evaluator.py:461]: 返回的game_segments是完整的几局轨迹吗？

File: zoo/classic_control/mountain_car/envs/mtcar_lightzero_env.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+@ENV_REGISTRY.register('mountain_car_lightzero')
+class MountainCarEnv(BaseEnv):
+    """
+    Implementation of LightZero's version of the Mountain Car deterministic MDP. 
+
+    Important references that contributed to the creation of this env:
+    > Source code of OpenAI's mountain car gym : https://is.gd/y1FkMT
+    > Gym documentation of mountain car : https://is.gd/29S0dt
+    > Based off LightZero existing implementation of cartpole_env.py
+    > LightZero's env creation conventions : https://is.gd/ZHLISj
```
[the remark this answers is not in the document]
Comment (puyuan1996) [reply zoo/classic_control/mountain_car/envs/mtcar_lightzero_env.py:18]: 需要写一下 
- 任务的整体介绍
- obs action reward episode_length_limit 等MDP信息的详细注释，

File: zoo/classic_control/mountain_car/config/mtcar_muzero_config.py (hunk 7 above)
Comment (puyuan1996) [reply zoo/classic_control/mountain_car/config/mtcar_muzero_config.py:90]: - 目前有到-110说明是能够达到终点的，需要看看目前设置下后期性能不稳定的原因，分析一下collect/eval/learn tb，梯度norm，输出policy logits等的变化情况。
- 可能的测试config: 
  - reanalyze_ratio = 1
  - grad_clip_value=0.5

File: lzero/worker/muzero_evaluator.py (hunk 8 above)
Comment (nighood) [reply lzero/worker/muzero_evaluator.py:461]: 一局

File: zoo/classic_control/mountain_car/envs/mtcar_lightzero_env.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (nighood) [reply zoo/classic_control/mountain_car/envs/mtcar_lightzero_env.py:18]: 我直接加mountain_car的gym链接在ipynb可以吗
https://www.gymlibrary.dev/environments/classic_control/mountain_car/
